airbyte-integrations/bases/standard-source-test/src/main/java/io/airbyte/integrations/standardtest/source/TestSource.java

```diff
@@ -225,22 +243,100 @@ public void testRead() throws Exception {
+    }
+
+    ConfiguredAirbyteCatalog configuredAirbyteCatalog = withSourceDefinedCursors(getConfiguredCatalog());
+    List<AirbyteMessage> airbyteMessages = runRead(configuredAirbyteCatalog, getState());
+    List<AirbyteRecordMessage> recordMessages = filterRecords(airbyteMessages);
+    List<AirbyteStateMessage> stateMessages = airbyteMessages
+        .stream()
+        .filter(m -> m.getType() == Type.STATE)
+        .map(AirbyteMessage::getState)
+        .collect(Collectors.toList());
+
+    assertFalse(recordMessages.isEmpty());
+    assertFalse(stateMessages.isEmpty());
+    // TODO validate exact records
+
+    // when we run incremental sync again there should be no new records. Run a sync with the latest
+    // state message and assert no records were emitted.
+    JsonNode latestState = stateMessages.get(stateMessages.size() - 1).getData();
+    List<AirbyteRecordMessage> secondSyncRecords = filterRecords(runRead(configuredAirbyteCatalog, latestState));
+    assertTrue(secondSyncRecords.isEmpty());
+  }
+
+  @Test
+  public void testEmptyStateIncrementalIdenticalToFullRefresh() throws Exception {
+    if (!sourceSupportsIncremental()) {
+      return;
+    }
+
+    ConfiguredAirbyteCatalog configuredCatalog = getConfiguredCatalog();
+    ConfiguredAirbyteCatalog fullRefreshCatalog = withFullRefreshSyncModes(configuredCatalog);
+
+    final List<AirbyteRecordMessage> fullRefreshRecords = filterRecords(runRead(fullRefreshCatalog));
+    final List<AirbyteRecordMessage> emptyStateRecords = filterRecords(runRead(configuredCatalog, Jsons.jsonNode(new HashMap<>())));
+
+    assertFalse(fullRefreshRecords.isEmpty());
+    assertFalse(emptyStateRecords.isEmpty());
+    assertSameRecords(fullRefreshRecords, emptyStateRecords);
+  }
+
+  private List<AirbyteRecordMessage> filterRecords(Collection<AirbyteMessage> messages) {
+    return messages.stream()
+        .filter(m -> m.getType() == Type.RECORD)
+        .map(AirbyteMessage::getRecord)
+        .collect(Collectors.toList());
+  }
+
+  private ConfiguredAirbyteCatalog withSourceDefinedCursors(ConfiguredAirbyteCatalog catalog) {
+    ConfiguredAirbyteCatalog clone = Jsons.clone(catalog);
+    for (ConfiguredAirbyteStream configuredStream : clone.getStreams()) {
+      if (configuredStream.getSyncMode() == INCREMENTAL && configuredStream.getStream().getSourceDefinedCursor()) {
+        configuredStream.setCursorField(configuredStream.getStream().getDefaultCursorField());
+      }
+    }
+    return clone;
+  }
+
+  private ConfiguredAirbyteCatalog withFullRefreshSyncModes(ConfiguredAirbyteCatalog catalog) {
+    ConfiguredAirbyteCatalog clone = Jsons.clone(catalog);
+    for (ConfiguredAirbyteStream configuredStream : clone.getStreams()) {
+      if (configuredStream.getStream().getSupportedSyncModes().contains(FULL_REFRESH)) {
+        configuredStream.setSyncMode(FULL_REFRESH);
+      }
+    }
+    return clone;
+  }
+
+  private boolean sourceSupportsIncremental() throws Exception {
```

Comment:
hahaha. TooClever™ indeed. i'm okay with this for now. @ChristopheDuong is adding a `supportIncremental` flag to the spec.json, so maybe we can use that instead when it is done.